gdnative-core/src/export/signal.rs

```diff
@@ -0,0 +1,85 @@
+use crate::core_types::{GodotString, Variant, VariantType};
+use crate::export::{ClassBuilder, ExportInfo, NativeClass, PropertyUsage};
+
+pub struct SignalBuilder<'a, C> {
+    class_builder: &'a ClassBuilder<C>,
+    name: GodotString,
+    args: Vec<SignalArgument>,
+}
+
+impl<'a, C: NativeClass> SignalBuilder<'a, C> {
+    pub(super) fn new(class_builder: &'a ClassBuilder<C>, name: &str) -> Self {
+        Self {
+            class_builder,
+            name: GodotString::from(name),
+            args: vec![],
+        }
+    }
+
+    /// Add an argument for the signal.
+    #[inline]
+    pub fn arg(mut self, argument: SignalArgument) -> Self {
+        self.args.push(argument);
+        self
+    }
+
+    /// Finish registering the signal.
+    #[inline]
+    pub fn done(self) {
+        self.class_builder.add_signal(Signal {
+            name: self.name,
+            args: self.args,
+        });
+    }
+}
+
+pub(crate) struct Signal {
+    pub name: GodotString,
+    pub args: Vec<SignalArgument>,
+}
+
+pub struct SignalArgument {
+    pub name: GodotString,
+    pub default: Variant,
+    pub export_info: ExportInfo,
+    pub usage: PropertyUsage,
+}
+
+/// Argument (or parameter) in a signal declaration.
+/// You can use one of the constructors for convenience, or provide each field separately.
+impl SignalArgument {
+    /// The most common way of constructing a signal argument -- name and type.
+    #[inline]
+    pub fn new(name: impl Into<GodotString>, argument_type: VariantType) -> Self {
+        Self {
+            name: name.into(),
+            default: Variant::nil(),
+            export_info: ExportInfo::new(argument_type),
+            usage: PropertyUsage::DEFAULT,
+        }
+    }
+
+    /// Construct a signal argument that accepts any dynamic type.
+    #[inline]
+    pub fn untyped(name: impl Into<GodotString>) -> Self {
+        Self {
+            name: name.into(),
+            default: Variant::nil(),
+            export_info: ExportInfo::new(VariantType::Nil),
+            usage: PropertyUsage::DEFAULT,
```

Comment:
I'm not sure if this works. I think you'd need `PropertyUsage::NIL_IS_VARIANT` (which is *not* exposed to GDNative) for this to work completely, but then again the argument types are not checked, so it might not matter anyway.

Reply:
Ah ok, we had this code in `impl NativeClass for FuncState`, so I assumed there's a special meaning with nil and default.

https://github.com/godot-rust/godot-rust/blob/99a0d343d42767c26d2bd1ebb51a3c5c9ae61989/gdnative-async/src/rt/func_state.rs#L34-L43

Do you remember what the intention was there?

Reply:
There was no intention. It was just necessary to provide a value for all the fields. Arbitrary values are ok for the `FuncState` signals, since the latter would likely never show up in the editor. I was just unsure what such values could mean for signals that might.

Reply:
I see. Then maybe we can go with `untyped()` which expresses the intention. If we find that another implementation fits better, we could change it transparently, and user code gets "auto-fixed".